### ISSUE TITLE
Segmentation Fault / ConnectionResetError: [WinError 10054] / Assertion failed: (exception_value) != NULL

### ISSUE THREAD

**Issue by jimkring** [Nuitka/Nuitka#3132]
I am having an issue where a my built app was crashing on Windows (`Nuitka: A segmentation fault has occurred.`). It seemed to happen when my application closed an http connection to a server.

Note that I tried with Python 3.12 and Python 3.11 on Windows -- same crash/problem.

### Debug Info

I rebuilt my app with `--debug` and then ran it with `--debugger`, which showed the following error when it crashed:

```
**Exception in callback _ProactorBasePipeTransport._call_connection_lost(None)
handle: <Handle _ProactorBasePipeTransport._call_connection_lost(None)>
Traceback (most recent call last):
  File "C:\Users\jim\AppData\Local\Temp\ONEFIL~2\asyncio\events.py", line 84, in _run
  File "C:\Users\jim\AppData\Local\Temp\ONEFIL~2\asyncio\proactor_events.py", line 165, in _call_connection_lost
ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host
Assertion failed: (exception_value) != NULL, file C:\Project\VENV~1\Lib\SITE-P~1\nuitka\build\include\nuitka/exceptions.h, line 1136**
```

### Nuitka Version

```
python -m nuitka --version  
2.5rc5
Commercial: 3.7.3
Python: 3.11.9 (tags/v3.11.9:de54cf5, Apr  2 2024, 10:12:12) [MSC v.1938 64 bit (AMD64)]
Flavor: CPython Official
Executable: C:\Project\.venv\Scripts\python.exe
OS: Windows
Arch: x86_64
WindowsRelease: 10
Version C compiler: C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.41.34120\bin\Hostx64\x64\cl.exe (cl 14.3).
```

### venv / packages

I'm using a virtual environment created with `uv venv` to do the build

```
PS C:\Project> uv pip list -v              
DEBUG uv 0.3.1
DEBUG Searching for Python interpreter in system path or `py` launcher
DEBUG Found `cpython-3.11.9-windows-x86_64-none` at `C:\Project\.venv\Scripts\python.exe` (active virtual environment)
DEBUG Using Python 3.11.9 environment at .venv\Scripts\python.exe
Package                       Version     Editable project location
----------------------------- ----------- -------------------------
annotated-types               0.7.0
anyio                         4.6.0
astroid                       3.3.4
beautifulsoup4                4.12.3
certifi                       2024.8.30
cfgv                          3.4.0
charset-normalizer            3.3.2
click                         8.1.7
colorama                      0.4.6
distlib                       0.3.8
distro                        1.9.0
docstring-parser              0.16
duckduckgo-search             6.2.13
fastapi                       0.115.0
filelock                      3.16.1
h11                           0.14.0
httpcore                      1.0.5
httpx                         0.27.2
identify                      2.6.1
idna                          3.10
importlib-metadata            8.5.0
iniconfig                     2.0.0
jaraco-classes                3.4.0
jinja2                        3.1.4
jiter                         0.5.0
keyring                       24.3.0
loguru                        0.7.2
markupsafe                    2.1.5
maturin                       1.7.4
mirascope                     1.2.0
more-itertools                10.5.0
nodeenv                       1.9.1
nuitka                        2.5rc5
openai                        1.50.2
ordered-set                   4.1.0
packaging                     24.1
pillow                        10.4.0
platformdirs                  4.3.6
pluggy                        1.5.0
pre-commit                    3.8.0
primp                         0.6.3
pydantic                      2.9.2
pydantic-core                 2.23.4
pynvml                        11.5.3
pytest                        8.3.3
python-dateutil               2.9.0.post0
python-dotenv                 1.0.1
pywin32-ctypes                0.2.3
pyyaml                        6.0.2
requests                      2.32.3
ruff                          0.6.8
screeninfo                    0.8.1
six                           1.16.0
sniffio                       1.3.1
soupsieve                     2.6
starlette                     0.38.6
tqdm                          4.66.5
typing-extensions             4.12.2
urllib3                       2.2.3
uvicorn                       0.31.0
virtualenv                    20.26.6
win32-setctime                1.1.0
zipp                          3.20.2
zstandard                     0.23.0
```

Thanks for your work making Nuitka awesome! :1st_place_medal:

**Comment by jimkring** [comment]
Also, I noticed a `debug.log` file with the following:
```
[0929/142230.428:ERROR:registration_protocol_win.cc(108)] CreateFile: The system cannot find the file specified. (0x2)
[0929/142230.932:ERROR:registration_protocol_win.cc(108)] CreateFile: The system cannot find the file specified. (0x2)
```

**Comment by kayhayen** [comment]
Something seems at odd with Nuitka and the assertion should never happen. I understand this one can be hard to reproduce, but I will try it.

Code is this:

```
    def _call_connection_lost(self, exc):
        if self._called_connection_lost:
            return
        try:
            self._protocol.connection_lost(exc)
        finally:
            # XXX If there is a pending overlapped read on the other
            # end then it may fail with ERROR_NETNAME_DELETED if we
            # just close our end.  First calling shutdown() seems to
            # cure it, but maybe using DisconnectEx() would be better.
            if hasattr(self._sock, 'shutdown') and self._sock.fileno() != -1:
                self._sock.shutdown(socket.SHUT_RDWR)
            self._sock.close()
            self._sock = None
            server = self._server
            if server is not None:
                server._detach()
                self._server = None
            self._called_connection_lost = True
```

The crash is on `self._sock.shutdown(socket.SHUT_RDWR)` in said finally block.

**Comment by kayhayen** [comment]
For the assertion, it complains about being called with a NULL checked. The fact that 3.11 and 3.12 behave the same is actually very important, due to the big change in how values are stored with 3.12 over 3.11 storing type, value, tb tripples. It would appear that coroutine code might be at fault, but said function is a normal function, so I don't know that, but `self._protocol.connection_lost(exc)` may well be that.

**Comment by kayhayen** [comment]
I have performed an audit, and found a few places, where Nuitka is relying on functions to set exceptions, that may don't do that, so I added assertions for exceptions to be set. If you re-run it with current staging branch, you would be supposed to get a better assertion.

I think, my CI will not find anything there, because the function if not assertion, surely has to segfault if it so happens to have the NULL value checked, so I think I need new input there. These debugging improvements however are always useful to have.
